### PR TITLE
[full-ci] new url format

### LINKED ISSUE #6085: New URL format
# Description

## User Stories

* > As a user receiving a link I would like to see in the URL what resource I'll access, so that I know beforehand that I don't open weird stuff.
* > As a user who keeps bookmarks in the browser, I'd like to bookmark the files I often work with, so that I can access them quickly. I want the bookmarks to still work, even when I change the folder name or move a folder to another location.
* > As a user I would like to be able to copy the URL and give it to a colleague, so that it's easier to collaborate quickly (side note: requires respective permissions to the resource e.g. via a group share or similar, and the resource id in the URL).

## Value
- The path makes it “human readable”
- The URL can be bookmarked
- The bookmarked URLs remain stable even if the path changes

## Acceptance Criteria
- URL has a fixed format until the relative path (which can, of course, have as many segments as needed)
- format is:
	- for views where we address a storage+path:
    	`http(s)://[...]/<app>/<storage-namespace>/<storage>/<relative path with slashes>?id=resourceid` (id query param is optional)
    - for views where we don't address a storage+path (e.g. favorites, which is a kind of meta view):
    	`http(s)://[...]/<app>/<view>`
    - this means that views that address a storage+path use a view-name which is the same as the respective storage namespace! (e.g. `personal` for personal view & personal storage namespace). For now this applies to `personal`, `shares` and `public`.
    - resource ids have the format base64(`<storage_id>:<file_id>`)
- resource loading:
	- path takes precedence over the id, but requires validation if both are given. workflow:
    - 1: try to load resources by path
    - 2: if loading by path fails:
    	- if id present: load by id + redirect to URL with correct path
        - if id absent: show "resource not found" page
    - 3: if loading by path succeeds:
		- if id present: must match with propfind response, otherwise "resource not found" page
		- if id absent: render result
- whether or not the URL contains an id as query param must be configurable (client side config, since it's not relevant for other client platforms)
- special case: `shared with me` will only be renamed to `shares`, don't start to implement the share jail story!
- vue router must be switched from `hash` to `history` mode -> also add documentation for webserver configs
	- history mode needs implementation for dev server and oCIS

## Definition of done
- Functional requirements
[ ] functionality described in the user story works
[ ] acceptance criteria are fulfilled
- Quality
[ ] code review happened
[ ] CI is green
[ ] critical code received unit tests by the developer
[ ] automated tests passed (if automated tests are not available, this test needs to be created and passed
- Non-functional requirements
[ ] no sonar cloud issues

### PR DESCRIPTION
**Pull request by fschade**
## Preface
CI will fail with integration tests, and maybe on e2e tests too.
It is meant as a preview to get test early and see what ci says.

The new route naming (we used string matching to identify routes) must be adopted in vendor applications.

## Description
this pr introduces the new URL design, it also touches some parts that are not discussed yet.
The technical design makes all url paths interchangeable without touching the logic.

To get a better background please read https://github.com/owncloud/web/issues/6085 and https://github.com/owncloud/ocis/blob/master/docs/ocis/adr/0011-global-url-format.md.

Following things are included:
* new url grouping (common, spaces, operations and shares)
* new files routes and route helpers
* refactoring of the breadcrumb to not rely on static pathes anymore
* route active detection now is based on vue router resolve instead of pure string matching
* preparations for spaces namespace and storage
* deprecation path for old routes

please read the routes overview section to get a better understanding of which parts have changed.

## Related Issue
* Fixes #6085

## Motivation and Context
prepare web to support spaces and introduce a clear mental cluster of what route should go where. 

## How Has This Been Tested?
* unit tests
* manual testing

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Technical debt
- [X] Tests

## Checklist:
- [X] Code changes
- [X] Unit tests added

## Open tasks:
- [X] implement route url schema
- [ ] enable history mode
- [ ] resource id - path fallback and error handling
- [ ] more manual testing
- [ ] remove dev leftovers

### Old schema
| name | path |
|---|---|
| files-personal | /files/list/all/:item* |
| files-favorites | /files/list/favorites |
| files-shared-with-me | /files/list/shared-with-me |
| files-shared-with-others | /files/list/shared-with-others |
| files-shared-via-link | /files/list/shared-via-link |
| files-trashbin | /files/list/trash-bin |
| files-public-list | /files/public/list/:item* |
| files-public-link | /files/public-link/:token |
| files-private-link | /files/private-link/:fileId |
| files-location-picker | /files/location-picker/:context/:action/:item* |
| files-public-drop | /files/files-drop/:token |

### New schema
| name | path |
|---|---|
| files-spaces-storage | /files/spaces/:type/:storage/:item* |
| files-shares-with-me | /files/shares/with-me |
| files-shares-with-others | /files/shares/with-others |
| files-shares-via-link | /files/shares/via-link |
| files-public-files | /files/public/show/:item* |
| files-public-drop | /files/public/drop/:token |
| files-common-favorites | /files/favorites |
| files-common-trash | /files/trash |
| files-operations-resolver-private-link | /files/ops/resolver/private-link/:fileId |
| files-operations-resolver-public-link | /files/ops/resolver/public-link/:token |
| files-operations-location-picker | /files/ops/location-picker/:context/:action/:item* |

#### /files/spaces
* **name:** -
* **example:** /files/spaces
* **description:** redirects to spaces-storage

#### /files/spaces/:namespace
* **name:** -
* **example:** /files/spaces/personal
* **description:** redirects to spaces-storage

#### /files/spaces/:namespace/:storage/:item*
* **name:** files-spaces-storage
* **example:** /files/spaces/personal/admin/any/file/path
* **description:**
    * if namespace is not given it falls back to `personal`
    * if storage is not given it falls back to `current user id`
    * for now namespace is hard coded, the implementation is prepared to request it from the server

#### /files/shares
* **name:** -
* **example:** /files/shares/public/files/foo/bar/baz
* **redirects:** files-shares-with-me

#### /files/public/show/:item*
* **name:** files-public-files
* **example:** /files/public/show/foo/bar/baz

#### /files/public/drop/:token
* **name:** files-public-drop
* **example:** /files/public/drop/45dfgdf5gs5f4g56gd

#### /files/shares/with-me
* **name:** files-shares-with-me
* **example:** /files/shares/with-me

#### /files/shares/with-others
* **name:** files-shares-with-others
* **example:** /files/shares/with-others

#### /files/shares/via-link
* **name:** files-shares-via-link
* **example:** /files/shares/via-link

#### /files/trash
* **name:** files-common-trash
* **example:** /files/trash

#### /files/favorites
* **name:** files-common-favorites
* **example:** /files/favorites

#### /files/ops/location-picker/:context/:action/:item*
* **name:** files-common-favorites
* **example:** /files/ops/location-picker/private/copy/a/deep/path

#### /files/ops/resolver/public-link/:token
* **name:** files-operations-resolver-public-link
* **example:** /files/ops/resolver/public-link/45dfgdf5gs5f4g56gd

#### /files/ops/resolver/private-link/:fileId
* **name:** files-operations-resolver-private-link
* **example:** /files/ops/resolver/private-link/45dfgdf5gs5f4g56gd

#### /files/list
* **name:** -
* **example:** /files/list
* **redirect:** spaces-storage (defaults namespace and storage)

#### /files/list/all/item*
* **name:** files-personal
* **example:** /files/list/all/a/deep/path
* **redirect:** spaces-storage (defaults namespace and storage)

#### /files/list/favorites
* **name:** files-favorites
* **example:** /files/list/favorites
* **redirect:** files-common-favorites

#### /files/list/shared-with-me
* **name:** files-shared-with-me
* **example:** /files/list/shared-with-me
* **redirect:** files-shares-with-me

#### /files/list/shared-with-others
* **name:** files-shared-with-others
* **example:** /files/list/shared-with-others
* **redirect:** files-shares-with-others

#### /files/list/shared-via-link
* **name:** files-shared-via-link
* **example:** /files/list/shared-via-link
* **redirect:** files-shares-via-link

#### /files/list/trash-bin
* **name:** files-trashbin
* **example:** /files/list/trash-bin
* **redirect:** files-common-trash

#### /files/public/list/:item*
* **name:** files-public-list
* **example:** /files/public/list/:item*
* **redirect:** files-public-files

#### /files/public-link/:token
* **name:** files-public-list
* **example:** /files/public-link/:token
* **redirect:** files-operations-resolver-public-link

#### /files/private-link/:fileId
* **name:** files-private-link
* **example:** /files/private-link/:fileId
* **redirect:** files-operations-resolver-private-link

#### /files/location-picker/:context/:action/:item*
* **name:** files-location-picker
* **example:** /files/location-picker/private/move/foo/
* **redirect:** -

#### /files/files-drop/:token
* **name:** files-public-drop
* **example:** /files/files-drop/5gfd4s56dfgf56gdf
* **redirect:** -

### CODE REVIEW

File: packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue [hunk 1]
```diff
@@ -49,6 +49,7 @@
           <translate>Password protected</translate>
         </oc-tag>
       </div>
+      test
```
Comment (pascalwengerter) [inline packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue:52]: @fschade 👀

File: packages/web-app-files/src/components/SideBar/Shares/Collaborators/ShowCollaborator.vue [hunk 2]
```diff
@@ -306,12 +307,12 @@ export default {
     },
 
     viaRouterParams() {
-      return {
-        name: 'files-personal',
+      // todo
```
Comment (pascalwengerter) [inline packages/web-app-files/src/components/SideBar/Shares/Collaborators/ShowCollaborator.vue:310]: Do...what?:)

File: packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue [hunk 3]
```diff
@@ -117,16 +119,15 @@ export default {
 
     viaRouterParams() {
       const viaPath = this.link.path
-
-      return {
-        name: 'files-personal',
+      // toDo
```
Comment (pascalwengerter) [inline packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue:122]: Do...what?:)

File: packages/web-app-files/src/index.js [hunk 4]
```diff
@@ -23,65 +35,72 @@ const appInfo = {
 }
 const navItems = [
   {
-    name: $gettext('All files'),
+    name: $gettext(`All files`),
```
Comment (pascalwengerter) [inline packages/web-app-files/src/index.js:38]: What motivated this change here?

**Comment by ScharfViktor** [comment]
### New schema
I don't know how deep you have to dig here. I checked only opening new url

| name | path | test result | 
|---|---|---|
| files-spaces-storage | /files/spaces/:namespace/:storage/:item* | ✔️ |
| files-shares-public-files | /files/shares/public/files/:item* |  |
| files-shares-public-link | /files/shares/public/link/:token | ❌#/files/public-link/ifpB942A74lZjRx |
| files-shares-public-drop | /files/shares/public/drop/:token | ❌/#/files/public-link/eOkmMjffDbsAwiY  |
| files-shares-private-link | /files/shares/private/link/:fileId | ✔️ |
| files-shares-with-me | /files/shares/with-me | ✔️ |
| files-shares-with-others | /files/shares/with-others | ✔️ |
| files-shares-via-link | /files/shares/via-link | ✔️ |
| files-common-favorites | /files/favorites | ✔️ |
| files-common-trash | /files/trash | ✔️ |
| files-operations-location-picker | /files/ops/location-picker/:context/:action/:item* | |


public-link and public-drop require logins

**Comment by fschade** [comment]
@ScharfViktor many thanks for testing and find the broken pathes, i missed a redirect from old to the new route. this is done now and can be re-tested.

**Comment by ScharfViktor** [comment]
I tried to run smoke tests:

This needs to change:
`await page.click('a[href="#/files/spaces/"]')` -> https://github.com/owncloud/web/blob/master/tests/smoke/support/page/files/allFiles.ts#L15

`await page.click('a[href="#/files/shares/with-me"]')` -> https://github.com/owncloud/web/blob/master/tests/smoke/support/page/files/sharedWithMe.ts#L12



But the tests were failed because of this problem (MOVE or COPY buttons were not found)

https://user-images.githubusercontent.com/84779829/146207760-410a83af-f01e-4ad4-89d7-e3a4a74e6867.mov

**Comment by fschade** [comment]
> I tried to run smoke tests:
> 
> This needs to change: `await page.click('a[href="#/files/spaces/"]')` -> https://github.com/owncloud/web/blob/master/tests/smoke/support/page/files/allFiles.ts#L15
> 
> `await page.click('a[href="#/files/shares/with-me"]')` -> https://github.com/owncloud/web/blob/master/tests/smoke/support/page/files/sharedWithMe.ts#L12
> 
> But the tests were failed because of this problem (MOVE or COPY buttons were not found)
> 
>  Untitled.mov

thank you for digging into it, the move, copy etc. actions where missed du the problem that web right now is not able to wait with rendering till the user is fetched (only right after the first login), that is known and as a temporary workaround i removed the user name from storage (/files/spaces/personal/alice) and use home instead (/files/spaces/personal/home).

Thats not cool but the only workaround i could think of till we update the boot mechanism to support those cases.

I think i catched all other problems too, lets see what ci is saying now :)

File: packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue (hunk 1 above)
Comment (fschade) [reply packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue]: dev leftover :)

File: packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue (hunk 3 above)
Comment (fschade) [reply packages/web-app-files/src/components/SideBar/Links/PublicLinks/LinkInfo.vue:121]: thanks, i already did and forgot to remove the comment :)

File: packages/web-app-files/src/index.js (hunk 4 above)
Comment (fschade) [reply packages/web-app-files/src/index.js:38]: good catch, i had dynamic variables in there while refactoring. I removed it

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingAccept</strong> https://drone.owncloud.com/owncloud/web/21096/15/1
The following scenarios passed on retry:
- webUISharingAcceptSharesToRoot/acceptShares.feature:108

**Review by tbsbdr — COMMENTED** [review]
as discussed: lgtm

File: packages/web-app-files/src/router/utils.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    return isLocationActive(router, first, ...rest)
+  }
+}
+
+/**
+ * helper to check if route needs authentication or not
+ *
+ * @param route
+ */
+export const isRoutePublic = (route: { meta?: RouteMeta }): boolean => {
```
Comment (kulmann) [inline packages/web-app-files/src/router/utils.ts:59]: I'd like to rename this to `isAuthRequired` or `requiresAuth` or something alike. IMO the word `public` is a reason for confusion, since we also have public links (completely different thing). What do you think?

File: packages/web-app-files/src/router/deprecated.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import { createLocationShares } from './shares'
+import { createLocationCommon } from './common'
+
+/**
+ * all route configs created by buildRoutes are deprecated,
+ * this helper wraps a route config and warns the user that it will go away and redirect to the new location.
+ *
+ * @param routeConfig
+ */
+const deprecatedRedirect = (routeConfig: {
```
Comment (kulmann) [inline packages/web-app-files/src/router/deprecated.ts:12]: Thank you for going this extra mile! Awesome!

File: packages/web-app-files/src/index.js [hunk 7]
```diff
@@ -26,62 +38,69 @@ const navItems = [
...
+    route: {
+      path: `/${appInfo.id}/trash`
+    },
     enabled(capabilities) {
       return capabilities.dav && capabilities.dav.trashbin === '1.0'
-    },
+    }
+  },
+  {
+    name: $gettext('Favorites'),
```
Comment (kulmann) [inline packages/web-app-files/src/index.js:76]: Why did you move Favorites to the end? Please move back to original order. 😉

**Review by kulmann — CHANGES_REQUESTED** [review]
Thanks for the heavy lifting and attention to detail. I have some things that we should discuss:
- the PR lacks a changelog item. As of now it needs to be a `change` as existing applications need to adapt (also see below)
- can we have the route names in constants? It feels repetitive that they are all over the place as strings
- could you think of a mapping of old route names to new route names as a kind of compatibility layer? I think it is needed in the `fileActions` mixin only (line 89). With that this PR could be an `enhancement` instead of a `change` afaict
- a note: I like that `isLocationSpacesActive(this.$router)` checks with the default route, but I don't like the readability of it when placed side by side with explicit route checks. I'd be in favour of making the check explicit as well ( `files-spaces-storage-personal`, I know that we don't have it but the readability would be much better).
- could you explain/remind me again why the personal files have to be in the `spaces` namespace? Technically shares are spaces as well and they don't live in the `spaces` namespace either. From a user perspective I'd expect to have `files/personal/home` and later on when project spaces are alive `files/spaces/marketing` and the like. `files/spaces/personal/home` is confusing to me.

(Some more smaller things in the comments)

File: packages/web-app-files/src/index.js (hunk 7 above)
Comment (fschade) [reply packages/web-app-files/src/index.js:76]: Shiddo, this happened due refactoring where I duplicated and relocated the links. Good catch

File: packages/web-app-files/src/router/deprecated.ts (hunk 6 above)
Comment (fschade) [reply packages/web-app-files/src/router/deprecated.ts:12]: Sure :)

File: packages/web-app-files/src/router/utils.ts (hunk 5 above)
Comment (fschade) [reply packages/web-app-files/src/router/utils.ts:59]: Yup make sense

**Comment by fschade** [comment]
> Thanks for the heavy lifting and attention to detail. I have some things that we should discuss:
> 
> * the PR lacks a changelog item. As of now it needs to be a `change` as existing applications need to adapt (also see below)
> * can we have the route names in constants? It feels repetitive that they are all over the place as strings
> * could you think of a mapping of old route names to new route names as a kind of compatibility layer? I think it is needed in the `fileActions` mixin only (line 89). With that this PR could be an `enhancement` instead of a `change` afaict
> * a note: I like that `isLocationSpacesActive(this.$router)` checks with the default route, but I don't like the readability of it when placed side by side with explicit route checks. I'd be in favour of making the check explicit as well ( `files-spaces-storage-personal`, I know that we don't have it but the readability would be much better).
> * could you explain/remind me again why the personal files have to be in the `spaces` namespace? Technically shares are spaces as well and they don't live in the `spaces` namespace either. From a user perspective I'd expect to have `files/personal/home` and later on when project spaces are alive `files/spaces/marketing` and the like. `files/spaces/personal/home` is confusing to me.
> * `files/shares/public/files/uGtt0VAc7hyu9Ts` looks repetitive as well. Would be much nicer to have just `files/shares/public/uGtt0VAc7hyu9Ts` would look better (without the second `files` segment).
> 
> (Some more smaller things in the comments)

@kulmann i hope I haven't missed a point on your list, please have a look. As we had additional feedback which could break things we need to re-run all tests did before. Cc.: @ScharfViktor

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingBasic</strong> https://drone.owncloud.com/owncloud/web/21475/53/1
The following scenarios passed on retry:
- webUISharingAcceptShares/acceptShares.feature:94

**Review by kulmann — APPROVED** [review]
Hooooray! 🥳 🎈 💯 🥇 Awesome piece of work, thank you @fschade ! ❤️

**Comment by kulmann** [comment]
Last CI run was green: https://drone.owncloud.com/owncloud/web/21475
Don't know why the status was not reported back. I'm fine with force merging if you give your GO.
